File: mobile/lib/utils/logger.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:smart_attendance_app/core/constants.dart';

/// Unified logger for the Smart Attendance mobile app.
///
/// Every call:
///  1. Prints to the debug console via [debugPrint].
///  2. Ships the event to the backend [kApiBaseUrl]/logs endpoint so it lands
///     in logs/mobile.log on the server.
///
/// All log payloads automatically include the app version and platform name.
/// A single retry is attempted on network timeout to handle brief connectivity
/// gaps without blocking the UI.
class AppLogger {
  AppLogger._(); // prevent instantiation

  static String? _userId;

  /// Declare the app version inline — kept in sync with pubspec.yaml version field.
  static const String _appVersion = '1.0.0+1';

  /// Set the authenticated user's ID so it is included in subsequent logs.
  static void setUserId(String? userId) {
    _userId = userId;
  }

  // --------------------------------------------------------------------------
  // Public log methods
  // --------------------------------------------------------------------------

  /// Fine-grained diagnostic information.
  static void debug(String message, {Map<String, dynamic>? context}) {
    _emit('DEBUG', message, context: context);
  }

  /// Routine operational events — navigation, successful API calls, etc.
  static void info(String message, {Map<String, dynamic>? context}) {
    _emit('INFO', message, context: context);
  }

  /// Non-fatal issues the system can recover from.
  static void warn(String message, {Map<String, dynamic>? context}) {
    _emit('WARN', message, context: context);
  }

  /// Errors that degrade user-visible functionality.
  static void error(String message, {Map<String, dynamic>? context}) {
    _emit('ERROR', message, context: context);
  }

  /// Severe failures — unhandled exceptions, app crashes.
  static void critical(String message, {Map<String, dynamic>? context}) {
    _emit('CRITICAL', message, context: context);
  }

  // --------------------------------------------------------------------------
  // Internal helpers
  // --------------------------------------------------------------------------

  static void _emit(
    String level,
    String message, {
    Map<String, dynamic>? context,
  }) {
    final prefix = '[${level.padRight(8)}]';
    debugPrint('$prefix $message');

    // Fire-and-forget — must not block the caller
    _sendLog(level, message, context: context);
  }

  static Future<void> _sendLog(
    String level,
    String message, {
    Map<String, dynamic>? context,
    int attempt = 0,
  }) async {
    final payload = <String, dynamic>{
      'source': 'mobile',
      'level': level,
      'message': message,
      'timestamp': DateTime.now().toUtc().toIso8601String(),
      'platform_version': _appVersion,
      'platform': _resolvePlatformName(),
      if (_userId != null) 'user_id': _userId,
      if (context != null) 'context': context,
    };

    try {
      await http
          .post(
            Uri.parse('$kApiBaseUrl/logs'),
            headers: {'Content-Type': 'application/json'},
            body: jsonEncode(payload),
          )
          .timeout(const Duration(seconds: 4));
    } on SocketException catch (e) {
      // Network unavailable — silently drop, no retry needed
      debugPrint('[AppLogger] Network unavailable, log dropped: $e');
    } on TimeoutException catch (e) {
      if (attempt < 1) {
        // One retry on timeout
        debugPrint('[AppLogger] Timeout, retrying log: $e');
        await _sendLog(level, message, context: context, attempt: attempt + 1);
      } else {
        debugPrint('[AppLogger] Retry failed, log dropped: $e');
      }
    } catch (e) {
      debugPrint('[AppLogger] Failed to send log: $e');
    }
  }

  static String _resolvePlatformName() {
    if (kIsWeb) return 'web';
    if (Platform.isAndroid) return 'android';
    if (Platform.isIOS) return 'ios';
    if (Platform.isWindows) return 'windows';
    if (Platform.isMacOS) return 'macos';
    if (Platform.isLinux) return 'linux';
    return 'unknown';
  }
}

File: frontend/src/utils/logger.ts
/**
 * Unified frontend logger for the Smart Attendance System.
 *
 * Every log call:
 *  1. Writes to the browser console for local development visibility.
 *  2. Ships the event to the backend /api/v1/logs endpoint so it lands in
 *     logs/frontend.log on the server.
 *
 * Log events are batched for 500 ms and sent in a single request to avoid
 * flooding the backend when many events fire in rapid succession (e.g. React
 * renders in strict mode).
 */

type LogLevel = 'DEBUG' | 'INFO' | 'WARN' | 'ERROR' | 'CRITICAL';

interface LogPayload {
  source: 'frontend';
  level: LogLevel;
  message: string;
  timestamp: string;
  context?: unknown;
  user_id?: string | null;
  platform_version?: string;
}

// Read the base URL once at module load time to avoid repeated env lookups.
const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

// App version sourced from the environment (set via next.config.ts if needed).
const APP_VERSION =
  process.env.NEXT_PUBLIC_APP_VERSION || 'unknown';

// ---------------------------------------------------------------------------
// Batching queue — events accumulate for BATCH_DELAY_MS then flush together
// ---------------------------------------------------------------------------
const BATCH_DELAY_MS = 500;
let _queue: LogPayload[] = [];
let _batchTimer: ReturnType<typeof setTimeout> | null = null;

function _scheduleFlush(): void {
  if (_batchTimer !== null) return;
  _batchTimer = setTimeout(() => {
    _batchTimer = null;
    _flush();
  }, BATCH_DELAY_MS);
}

async function _flush(): Promise<void> {
  if (_queue.length === 0) return;
  const batch = _queue.splice(0, _queue.length);

  for (const payload of batch) {
    try {
      await fetch(`${API_BASE_URL}/logs`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        // Keep-alive ensures the request completes even during page unload
        keepalive: true,
      });
    } catch {
      // Silently swallow — we must never let logger failures affect the UI
    }
  }
}

// ---------------------------------------------------------------------------
// Resolve the current user ID from localStorage (set by the auth store)
// ---------------------------------------------------------------------------
function _resolveUserId(): string | null {
  if (typeof window === 'undefined') return null;
  try {
    const raw = localStorage.getItem('auth_user');
    if (!raw) return null;
    const parsed = JSON.parse(raw) as { id?: string };
    return parsed?.id ?? null;
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Core send helper
// ---------------------------------------------------------------------------
function _enqueue(level: LogLevel, message: string, context?: unknown): void {
  // Never send logs during test runs
  if (process.env.NODE_ENV === 'test') return;

  const payload: LogPayload = {
    source: 'frontend',
    level,
    message,
    timestamp: new Date().toISOString(),
    platform_version: APP_VERSION,
    user_id: _resolveUserId(),
    ...(context !== undefined && context !== null ? { context } : {}),
  };

  _queue.push(payload);
  _scheduleFlush();
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------
export const AppLogger = {
  /** Fine-grained diagnostic information — not written in production unless DEBUG mode is on. */
  debug(message: string, context?: unknown): void {
    if (process.env.NODE_ENV === 'development') {
      // eslint-disable-next-line no-console
      console.debug(`[DEBUG] ${message}`, context ?? '');
    }
    _enqueue('DEBUG', message, context);
  },

  /** Routine operational events — page views, successful API calls, etc. */
  info(message: string, context?: unknown): void {
    // eslint-disable-next-line no-console
    console.info(`[INFO] ${message}`, context ?? '');
    _enqueue('INFO', message, context);
  },

  /** Non-fatal issues that the system can recover from. */
  warn(message: string, context?: unknown): void {
    // eslint-disable-next-line no-console
    console.warn(`[WARN] ${message}`, context ?? '');
    _enqueue('WARN', message, context);
  },

  /** Errors that affect user-visible functionality. */
  error(message: string, context?: unknown): void {
    // eslint-disable-next-line no-console
    console.error(`[ERROR] ${message}`, context ?? '');
    _enqueue('ERROR', message, context);
  },

  /** Severe errors — unhandled exceptions, application crashes. */
  critical(message: string, context?: unknown): void {
    // eslint-disable-next-line no-console
    console.error(`[CRITICAL] ${message}`, context ?? '');
    _enqueue('CRITICAL', message, context);
  },
};
